tests/integration/test_zookeeper_session_lifetime/test.py

```diff
@@ -0,0 +1,111 @@
+from helpers.network import PartitionManager
+
+cluster = ClickHouseCluster(
+    __file__, zookeeper_config_path="configs/zookeeper_load_balancing.xml"
+)
+
+node1 = cluster.add_instance(
+    "node1",
+    with_zookeeper=True,
+    main_configs=["configs/remote_servers.xml", "configs/zookeeper_load_balancing.xml"],
+)
+node2 = cluster.add_instance(
+    "node2",
+    with_zookeeper=True,
+    main_configs=["configs/remote_servers.xml", "configs/zookeeper_load_balancing.xml"],
+)
+node3 = cluster.add_instance(
+    "node3",
+    with_zookeeper=True,
+    main_configs=["configs/remote_servers.xml", "configs/zookeeper_load_balancing.xml"],
+)
+
+
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        cluster.start()
+        for node in [node1, node2, node3]:
+            node.query("DROP TABLE IF EXISTS simple SYNC")
+            node.query(
+                """
+            CREATE TABLE simple (date Date, id UInt32)
+            ENGINE = ReplicatedMergeTree('/clickhouse/tables/0/simple', '{replica}') PARTITION BY toYYYYMM(date) ORDER BY id;
+            """.format(
+                    replica=node.name
+                )
+            )
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+def assert_uses_zk_node(node: ClickHouseInstance, zk_node):
+    assert (
+        "1"
+        == str(
+            node.exec_in_container(
+                [
+                    "bash",
+                    "-c",
+                    "lsof -a -i4 -i6 -itcp -w | grep 'testzookeepersessionlifetime_{}_1.*testzookeepersessionlifetime_default:2181' | grep ESTABLISHED | wc -l".format(
+                        zk_node
+                    ),
+                ],
+                privileged=True,
+                user="root",
+            )
+        ).strip(),
+        "{} is not using zookeeper node {}".format(node.hostname, zk_node),
+    )
+
+
+def test_force_session_expiry(started_cluster: ClickHouseCluster):
+    pm = PartitionManager()
+
+    # only leave connecting to zoo3 possible
+    for node in started_cluster.instances.values():
+        for zk in ["zoo1", "zoo2"]:
+            pm._add_rule(
+                {
+                    "source": node.ip_address,
+                    "destination": cluster.get_instance_ip(zk),
+                    "action": "REJECT --reject-with tcp-reset",
+                }
+            )
+    try:
+        node1.query("INSERT INTO simple VALUES ({0}, {0})".format(1))
+
+        # give time for ZK sessions to time out because of the network partition
+        time.sleep(1)
+        for node in [node1, node2, node3]:
+            # all nodes will have to switch to zoo3
+            assert_uses_zk_node(node, "zoo3")
+
+        # and replication still works
+        for node in [node2, node3]:
+            assert node.query_with_retry("SELECT count() from simple") == "1\n"
+
+        pm.heal_all()
+
+        # wait for zk sessions to reach deadline
+        time.sleep(5)
```

Comment:
It's unreliable, the test is going to be flaky

Reply:
Any suggestion on how to improve? The test needs to wait for ZK sessions to expire and re-connect. They are set to expire between 2 and 4 seconds, hence 5. I understand sleeping for a magic number of seconds is not ideal.
Is there a way to query clickhouse itself for which ZK replica it's connected to? Maybe we could use that in `query_with_retry`

Reply:
Maybe `system.zookeeper_connection`, `query_with_retry` and `zookeeper_load_balancing` will help